### PR TITLE
QoS Errors

### PR DESCRIPTION
**Pull request by hestolz**
Patch fluentd `log.rb` in order to fork Error- and Fatal-level log events to the QoS pipeline. Log events of this level can then be tallied (in the case of repetitions) and then the top N=5 buckets will be uploaded.

Additional
- Updated telemetry to monitor OMI using new process name (now `dsc_host`)
- Eliminate chance of heavy duplication and hanging of omicli calls for telemetry due to omi hang by terminating calls after given deadline
- Tweak format of some existing log statements

### CODE REVIEW

File: source/code/plugins/in_dsc_monitor.rb [hunk 1]
```diff
@@ -94,7 +94,7 @@ def run_check_status
         sleep @check_status_interval
       end 
       rescue => e
-        $log.error e
+        $log.error "DSC check status failed. Error: #{e}"
```
[the remark this answers is not in the document]
Comment (hestolz) [reply source/code/plugins/in_dsc_monitor.rb]: whats the tangible benefit

File: source/ext/patches/fluentd/log.patch [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
++require_relative '../../../../../../../../../plugin/agent_telemetry_script'
+ 
+ module Fluent
+   class Log
+@@ -273,6 +274,21 @@
+         }
+         record['message'] = message.dup
+         @engine.push_log_event("#{@tag}.#{level}", time.to_i, record)
++
++        case Log.str_to_level(level.to_s)
```
[the remark this answers is not in the document]
Comment (hestolz) [reply source/ext/patches/fluentd/log.patch]: smart, I will test it out ;)

File: source/code/plugins/in_dsc_monitor.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (abenbachir) [reply source/code/plugins/in_dsc_monitor.rb]: "$log" means use the global logger, and @log mean use the local (plugin) logger.
log_level configuration for each plugin will affect this logger "@log".

The $log refer to global log_level can can only be controlled from "<system>" configuration. as below:
<system>
  log_level info
</system>

File: source/ext/patches/fluentd/log.patch [hunk 3]
```diff
@@ -0,0 +1,24 @@
+--- ../source/ext/fluentd/lib/fluent/log.rb	2018-11-01 14:01:03.000703433 -0700
++++ ../source/ext/fluentd/lib/fluent/log.rb.new	2019-05-30 15:06:52.974607146 -0700
+@@ -15,6 +15,7 @@
+ #
+ 
+ require 'forwardable'
++require_relative '../../../../../../../../../plugin/agent_telemetry_script'
```
[the remark this answers is not in the document]
Comment (hestolz) [reply source/ext/patches/fluentd/log.patch]: unfortunately this doens't work

File: source/code/plugins/in_dsc_monitor.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (hestolz) [reply source/code/plugins/in_dsc_monitor.rb]: if i do i'll send a separate PR for it. $ is being used in almost every plugin; after all most of the conf files that we install don't specify a log_level ...

File: source/code/plugins/in_agent_telemetry.rb [hunk 4]
```diff
@@ -72,8 +72,13 @@ def run_periodic
           @query_interval = query_interval if !query_interval.nil? and query_interval.between?(MIN_QUERY_INTERVAL, MAX_QUERY_INTERVAL)
           next_heartbeat = now + @query_interval
         end
-        @telemetry_script.poll_resource_usage
+        t = Thread.new { @telemetry_script.poll_resource_usage }
```
Comment (hestolz) [inline source/code/plugins/in_agent_telemetry.rb]: @abenbachir suggested tread changes

**Comment by abenbachir** [comment]
@hestolz, Can we merge this PR ?

**Pull request merged**